[MCP SDK] Handle closed connection in `StreamTransport` to avoid stuck worker

Gregory-Gerard opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Docs?         | no
| Issues        | Fix #283
| License       | MIT

WIP: I'll update this description with full details as I progress 👍

Gregory-Gerard commented:
@OskarStark would you be okay if I try the namespaced function override approach just for `connection_aborted` via a test-only polyfill? We can keep flush/write as-is and assert via PHP’s output buffering. 

I couldn’t find other example testing this; even HttpFoundation’s `EventStreamResponse` doesn’t check specifically `connection_aborted`. 

This would avoids the `Runtime` dependency and any constructor changes. If you’re fine with it, I’ll add a dev-only polyfill that shadows `Symfony\AI\McpSdk\Server\Transport\Sse\connection_aborted()` in tests.

OskarStark commented:
I would say give it a try, why not

Gregory-Gerard commented:
Re @OskarStark! I implemented the namespaced override only for `connection_aborted`. `StreamTransport` keeps native calls; tests assert via output capture; and no new `RuntimeInterface`. If you’re good with this, I’ll proceed with the concrete SSE ping/heartbeat fix.

OskarStark commented on src/mcp-sdk/tests/Server/Transport/Sse/StreamTransportTest.php:
```diff
@@ -0,0 +1,130 @@
+
+    public function testSendEmitsMessageEventAndFlushes()
+    {
+        $store = $this->createMock(StoreInterface::class);
+        $id = Uuid::v7();
```
Let's inline one time variables

Gregory-Gerard replied:
Done!

OskarStark commented:
Yes why not 🙌🙂

Gregory-Gerard commented:
Hi @OskarStark @chr-hertel (friendly tag because you were tagged in the issue originally)!

Quick update:

- Still WIP. Next up: make ping interval and request timeouts configurable per MCP spec (https://modelcontextprotocol.io/specification/2025-06-18/basic/utilities/ping). I may introduce a small immutable ServerConfig injected into `Server` to avoid constructor bloat.
- Ping (or rather the timeout) is non-strict for now because some clients don't reply explicitly to pings. I'll add a "strict" mode (default off) so we can enforce spec behavior without breaking common clients (spec: https://modelcontextprotocol.io/specification/2025-06-18/basic/lifecycle#timeouts, e.g. issue with a strict timeout https://github.com/punkpeye/fastmcp/issues/38).
- Major changes so far:
  - Added `PendingResponse`/`PendingResponseBag` to track outgoing requests, resolve inbound responses/errors, and GC on timeout. This currently requires `JsonRpcHandler` to yield `Response|Error`. Longer-term I'd prefer splitting JSON encode/decode from processing handlers as proposed in PR #240, but seems out of scope here, but happy to adjust later if you prefer.
  - Added `KeepAliveSessionInterface` + `KeepAliveSession` to automate periodic pings.
  - `Server::sendRequest(...)` with callback and timeout handling is in; next I'll add per-request timeout overrides and emit a `cancel` notification to the client on timeout (which implies adding `sendNotification`), still per the spec.
  - Injected a `ClockInterface` to make testing easier and remove a direct dependency on system clock with `usleep`

WDYT about my directions? Happy to have feedback before polishing thing!

Gregory-Gerard commented on src/mcp-sdk/src/Server.php:
```diff
@@ -13,22 +13,42 @@
+
     public function __construct(
         private JsonRpcHandler $jsonRpcHandler,
+        private ClockInterface $clock,
```
Thinking about adding a default `Clock` like in Symfony Messenger, wdyt?
